src/dune_rpc_server/dune_rpc_server.ml

```diff
@@ -1,5 +1,47 @@
 open Stdune
 open Dune_rpc_private
+open Fiber.O
+
+module Subscription = struct
+  module Id = Stdune.Id.Make ()
+
+  type 'a t =
+    { send : 'a option -> unit Fiber.t
+    ; finished : unit Fiber.Ivar.t
+    ; mutable active : bool
+    ; id : Id.t
+    }
+
+  let to_dyn t = Id.to_dyn t.id
+
+  type packed = E : _ t -> packed
+
+  let compare x y = Id.compare x.id y.id
+
+  let compare_packed (E x) (E y) = Id.compare x.id y.id
+
+  let finished t = Fiber.Ivar.read t.finished
+
+  let active t = t.active
+
+  let finish' t ~by_ =
+    match t.active with
+    | false -> Fiber.return ()
+    | true -> (
+      t.active <- false;
+      match by_ with
+      | `Client -> Fiber.Ivar.fill t.finished ()
+      | `Server ->
+        Fiber.fork_and_join_unit (Fiber.Ivar.fill t.finished) (fun () ->
+            t.send None))
+
+  let finish = finish' ~by_:`Server
```

Comment:
Should we be using `Fiber.of_thunk` here and in `update` to avoid running effects before fiber is executed?